akka-projection-dynamodb/src/main/scala/akka/projection/dynamodb/internal/DynamoDBProjectionImpl.scala

```diff
@@ -420,6 +422,90 @@ private[projection] object DynamoDBProjectionImpl {
+      sourceProvider: SourceProvider[Offset, Envelope],
+      offsetStore: DynamoDBOffsetStore,
+      originalEnvelope: Envelope,
+      handler: Handler[Envelope])(implicit ec: ExecutionContext, system: ActorSystem[_]): Future[Boolean] = {
+    originalEnvelope match {
+      case originalEventEnvelope: EventEnvelope[Any @unchecked] if originalEventEnvelope.sequenceNr > 1 =>
+        sourceProvider match {
+          // FIXME config to make this case opt in
+          case provider: LoadEventsByPersistenceIdSourceProvider[Any @unchecked] =>
+            val persistenceId = originalEventEnvelope.persistenceId
+            offsetStore.storedSeqNr(persistenceId).flatMap { storedSeqNr =>
+              val fromSeqNr = storedSeqNr + 1
+              val toSeqNr = originalEventEnvelope.sequenceNr
+              provider.currentEventsByPersistenceId(persistenceId, fromSeqNr, toSeqNr) match {
+                case Some(querySource) =>
+                  querySource
+                    .mapAsync(1) { envelope =>
+                      import DynamoDBOffsetStore.Validation._
+                      offsetStore
+                        .validate(envelope)
+                        .flatMap {
+                          case Accepted =>
+                            if (isFilteredEvent(envelope)) {
+                              offsetStore.addInflight(envelope)
+                              FutureDone
+                            } else {
+                              handler
+                                .process(envelope.asInstanceOf[Envelope])
+                                .map { _ =>
+                                  offsetStore.addInflight(envelope)
+                                  Done
+                                }
+                            }
+                          case Duplicate =>
+                            FutureDone
+                          case RejectedSeqNr =>
+                            // this shouldn't happen
+                            throw new RejectedEnvelope(
+                              s"Replay due to rejected envelope was rejected. PersistenceId [$persistenceId] seqNr [${envelope.sequenceNr}].")
+                          case RejectedBacktrackingSeqNr =>
+                            // this shouldn't happen
+                            throw new RejectedEnvelope(
+                              s"Replay due to rejected envelope was rejected. Should not be from backtracking. PersistenceId [$persistenceId] seqNr [${envelope.sequenceNr}].")
+                        }
+                    }
+                    .runFold(0) { case (acc, _) => acc + 1 }
+                    .map { count =>
+                      val expected = toSeqNr - fromSeqNr
+                      if (count == expected) {
+                        true
+                      } else {
+                        // it's expected to find all events, otherwise fail the replay attempt
+                        log.warn(
```

Comment:
do we have logPrefix here, or can we add that in same way as the offset store?

Reply:
Added the offset store log prefix.